metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -129,155 +149,402 @@ func getResource(resourceName string) kubernetes.Resource {
+	case DaemonSetResource:
+		return []string{NamespaceResource}
+	case JobResource:
+		return []string{NamespaceResource}
+	case CronJobResource:
+		return []string{NamespaceResource}
+	case PersistentVolumeResource:
+		return []string{}
+	case PersistentVolumeClaimResource:
+		return []string{NamespaceResource}
+	case StorageClassResource:
+		return []string{}
+	case NodeResource:
+		return []string{}
+	case NamespaceResource:
+		return []string{}
+	default:
+		return []string{}
+	}
+}
+
+// getWatchOptions builds the kubernetes.WatchOptions{} needed for the watcher based on the config and nodeScope
+func getWatchOptions(config *kubernetesConfig, nodeScope bool, client k8sclient.Interface, log *logp.Logger) (*kubernetes.WatchOptions, error) {
+	var err error
+	options := kubernetes.WatchOptions{
+		SyncTimeout: config.SyncPeriod,
+		Namespace:   config.Namespace,
+	}
+
+	// Watch objects in the node only
+	if nodeScope {
+		nd := &kubernetes.DiscoverKubernetesNodeParams{
+			ConfigHost:  config.Node,
+			Client:      client,
+			IsInCluster: kubernetes.IsInCluster(config.KubeConfig),
+			HostUtils:   &kubernetes.DefaultDiscoveryUtils{},
+		}
+		options.Node, err = kubernetes.DiscoverKubernetesNode(log, nd)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't discover kubernetes node: %w", err)
+		}
+	}
+	return &options, err
+}
+
+// startWatcher starts a watcher for a specific resource
+func startWatcher(
+	resourceName string,
+	resource kubernetes.Resource,
+	options kubernetes.WatchOptions,
+	client k8sclient.Interface,
+	resourceWatchers *Watchers) (bool, error) {
+
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
+
+	_, ok := resourceWatchers.watchersMap[resourceName]
+	// if it does not exist, create the watcher
+	if !ok {
+		watcher, err := kubernetes.NewNamedWatcher(resourceName, client, resource, options, nil)
+		if err != nil {
+			return false, err
+		}
+		resourceWatchers.watchersMap[resourceName] = &watcherData{watcher: watcher, started: false}
+		return true, nil
+	}
+	return false, nil
+}
+
+// addToWhichAreUsing adds resource identified by usingName to the list of resources using the watcher
+// identified by resourceName
+func addToWhichAreUsing(resourceName string, usingName string, resourceWatchers *Watchers) {
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
+
+	data, ok := resourceWatchers.watchersMap[resourceName]
+	if ok {
+		contains := false
+		for _, which := range data.whichAreUsing {
+			if which == usingName {
+				contains = true
+				break
+			}
+		}
+		// add this resource to the list of resources using it
+		if !contains {
+			data.whichAreUsing = append(data.whichAreUsing, usingName)
+		}
+	}
+}
+
+// removeFromWhichAreUsing returns true if element was removed and new size of array.
+// The cache should be locked when called.
+func removeFromWhichAreUsing(resourceName string, notUsingName string, resourceWatchers *Watchers) (bool, int) {
```

Comment:
My concern here is the following:

- Namespace watcher is started from pod metricset and state_pod metricset
- The resourceName in both cases is pod
- From `addToWhichAreUsing` I understood that if resource name is contained, it is not added again. So the array will have pod once
- Enricher stop is called from pod metricset. Pod resource is removed and if it was the only resource(with distinct name) using it, watcher gets stopped
- But state_pod still wants it

Reply:
You are right, there is a conflict happening now between:
- state_pod and pod
- state_container and container
- state_node and node
I will see which logic can be changed and I will add a unit test for this case

Reply:
I will save to the array of the ones using the metricset name instead of the resource name to avoid these conflicts

Reply:
In my latest commit, I fix this conflict. Using `state_pod` and `pod` has the following output:
```
{"log.level":"debug","@timestamp":"2024-01-26T14:11:19.270Z","log.logger":"kubernetes","log.origin":{"function":"github.com/elastic/beats/v7/metricbeat/module/kubernetes/util.createAllWatchers","file.name":"util/kubernetes.go","file.line":332},"message":"Created watcher pod successfully, created by pod.","service.name":"metricbeat","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2024-01-26T14:11:19.270Z","log.logger":"kubernetes","log.origin":{"function":"github.com/elastic/beats/v7/metricbeat/module/kubernetes/util.createAllWatchers","file.name":"util/kubernetes.go","file.line":346},"message":"Created watcher state_namespace successfully, created by pod.","service.name":"metricbeat","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2024-01-26T14:11:19.270Z","log.logger":"kubernetes","log.origin":{"function":"github.com/elastic/beats/v7/metricbeat/module/kubernetes/util.createAllWatchers","file.name":"util/kubernetes.go","file.line":346},"message":"Created watcher node successfully, created by pod.","service.name":"metricbeat","ecs.version":"1.6.0"}
```

I also added a new unit test for metricsets.

Reply:
This is the right approach, I agree. BTW the results you pasted do not show the fix because all watchers were created by pod metricset apparently.

Reply:
Yes, exactly. I don't think it is possible to see it through logs, but I added a unit test for this case.